Ui

cso7654 opened the pull request:
# Description

Added a new UI canvas prefab pre-set with a health bar, color indicator, and crosshair, all including relevant scripts/functionality and ready to be interacted with through other objects. Prefab is currently over the world object as a HUD

## Type of change

New Feature

## How were the changes tested

Tested UI panel in fullscreen, maximized, and preview mode with no undesirable results. Different test inputs were provided and used and did not break functionality

## Self-Review Checklist:

- [X] Code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings or errors

cso7654 committed:
Add files via upload

cso7654 committed:
Add files via upload

cso7654 committed:
Add files via upload

cso7654 committed:
Restored with GitHub Desktop

LegendRK commented on Digital 1/Assets/Scripts/UI/HUD/Crosshair.cs:
```diff
@@ -0,0 +1,79 @@
+    public float maxCharge;
+    public Color defaultColor;
+    public Color chargeColor;
+    public float retractionTime;
```
These should all be private, with some having [SerializeField] in case you are setting some of them in the inspector. The ones that are changed outside of the class should have properties for public accessibility.

EthanNichols commented on Digital 1/Assets/Scenes/EthanScene.unity:
```diff
@@ -339,6 +339,125 @@ MonoBehaviour:
   m_Name: 
```
You shouldn't be modifying my scene. You should have your own scene.

LegendRK commented on Digital 1/Assets/Scripts/UI/HUD/HealthBar.cs:
```diff
@@ -0,0 +1,72 @@
+public class HealthBar : MonoBehaviour
+{
+    public GameObject healthBarPanel;
+    public GameObject healthBarFill;
```
should be private with [SerializeField]

LegendRK commented on Digital 1/Assets/Scripts/UI/HUD/ColorIndicator.cs:
```diff
@@ -0,0 +1,73 @@
+            {
+                FillColor = testColors[1];
+            }
+        }
```
Make sure to remove this before merging if it's meant for testing.

LegendRK commented on Digital 1/Assets/Scripts/UI/HUD/ColorIndicator.cs:
```diff
@@ -0,0 +1,73 @@
+            fillColor = value;
+            UpdateFill();
+        }
+    }
```
Redundant to have public fill color with properties. Make fill Color private

EthanNichols commented on Digital 1/Assets/Scripts/UI/HUD/HealthBar.cs:
```diff
@@ -0,0 +1,72 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class HealthBar : MonoBehaviour
```
You should look into the Slider UI.
Allows you to set a min/max value as well as the fill value for the slider.
So you aren't reinventing the wheel.

cso7654 replied:
This is not what the Slider UI is for. The health bar is just a visual of a health value, not an input for health. Panels/images are the right way to go with this.

EthanNichols replied:
The slider ui is a "Slider" you can disable the intractable aspect of the UI.
It's meant for any UI that slides/fills.

EthanNichols commented on Digital 1/Assets/Scripts/UI/HUD/Crosshair.cs:
```diff
@@ -0,0 +1,79 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
```
add
`using UnityEngine.UI` since there are ui references in the script.

cso7654 committed:
Refactored scripts

Changed public variables to [SerializeField} private variables. Added properties to said variables. Removed UI from EthanScene.  Added more comments.

LegendRK reviewed (commented):
Some issues that I see with the current UI
- the health bar and color indicator cover up a large part of the screen. I would make the health bar smaller and narrower, and move it to a corner (maybe reduce the alpha). 
- For the color indicator, maybe change the color of the crosshair to free up room on the screen.
- There isn't any plans for "charging" so I would remove that for now (or at least comment it).

cso7654 committed:
Changed HUD Layout

Shrunk health bar and color indicator. Disabled charge on crosshair and made it transparent to increase visibility.

cso7654 commented:
Health bar and color indicator were shrunk and moved further into the corners. The crosshair is configurable to not use charge shot animations, so it is disabled but can be enabled later if our minds are changed. Please be sure the game is running in 16:9 aspect ratio, otherwise Unity's canvas can get confused, and 16:9 is the preferred final ratio. If we need it to be able to handle multiple ratios, let me know and I will implement it, but that will require a significant amount of work to remain consistent.

cso7654 committed:
Changed opacity of Health/Color indicators

from 255 to 150. Will add dynamic opacity later when needed.

cso7654 committed:
replaced health bar with slider

it's more complicated now but it still looks the same

jmortiger commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Starter UI](https://trello.com/c/Amv5EbMJ/19-starter-ui)
